File: core/src/test/java/be/geoffrey/xmlweave/core/usecase/XmlWeaveInteractorTest.java
package be.geoffrey.xmlweave.core.usecase;

import org.junit.Test;

import java.io.File;
import java.util.Optional;
import java.util.stream.Collectors;

import static org.assertj.core.api.Assertions.assertThat;

public class XmlWeaveInteractorTest {

    @Test
    public void chooseInitialElementWithNoOptionsShouldNotReturnAnything() {

        File testFile = TestFileReader.readTestFile("2_simple_elements.xsd");

        XmlWeaveInteractor sut = new XmlWeaveInteractor();
        Optional<Element> representation = sut.getRepresentation(testFile, "");
        assertThat(representation).isNotPresent();
    }

    @Test
    public void chooseInitialElementWithSelectionShouldRepresentRootElement() {

        File testFile = TestFileReader.readTestFile("2_simple_elements.xsd");

        XmlWeaveInteractor sut = new XmlWeaveInteractor();
        Optional<Element> representation = sut.getRepresentation(testFile, "SimpleBasicElement");

        assertThat(representation).isPresent();
        assertThat(representation.get().getName()).isEqualTo("SimpleBasicElement");
    }

    @Test
    public void chooseInitialElementWithNonExistingSelectionShouldReturnNothing() {

        File testFile = TestFileReader.readTestFile("2_simple_elements.xsd");

        XmlWeaveInteractor sut = new XmlWeaveInteractor();
        Optional<Element> representation = sut.getRepresentation(testFile, "Bla");

        assertThat(representation).isNotPresent();
    }

    @Test
    public void simpleChildElementsShouldBeRenderedCorrectly() {

        File testFile = TestFileReader.readTestFile("1_simple_element_with_basic_childs.xsd");

        XmlWeaveInteractor sut = new XmlWeaveInteractor();
        Optional<Element> rep = sut.getRepresentation(testFile, "SimpleBasicElement");

        Element element = rep.get();

        assertThat(element.getChildren()).hasSize(2);
        assertThat(element.getChildren()
                .stream()
                .map(Element::getName)
                .collect(Collectors.toList()))
                .containsExactly("elementOne", "elementTwo");
    }

}

File: core/src/test/java/be/geoffrey/xmlweave/core/usecase/XmlWeaveInteractorTest.kt
package be.geoffrey.xmlweave.core.usecase

import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import java.util.*

class XmlWeaveInteractorTest {

    @Test
    fun chooseInitialElementWithNoOptionsShouldNotReturnAnything() {
        val representation = interpretTestFile("2_simple_elements.xsd", "")
        assertThat(representation).isNotPresent
    }

    @Test
    fun chooseInitialElementWithSelectionShouldRepresentRootElement() {
        val representation = interpretTestFile("2_simple_elements.xsd", "SimpleBasicElement")
        assertThat(representation).isPresent
        assertThat(representation.get().name).isEqualTo("SimpleBasicElement")
    }

    @Test
    fun chooseInitialElementWithNonExistingSelectionShouldReturnNothing() {
        val representation = interpretTestFile("2_simple_elements.xsd", "Bla")
        assertThat(representation).isNotPresent
    }

    @Test
    fun simpleChildElementsShouldBeInterpretedCorrectly() {
        val rep = interpretTestFile("1_simple_element_with_local_complex_type.xsd", "SimpleBasicElement").get()

        assertThat(rep.children
                .map { it.name })
                .containsExactly("elementOne", "elementTwo")
    }

    @Test
    fun simpleComplexTypeShouldBeInterpreted() {
        val rep = interpretTestFile("3_simple_element_with_non_local_complex_type.xsd", "SimpleBasicElement").get()

        assertThat(rep.children
                .map { it.name })
                .containsExactly("elementOne", "elementTwo")
    }

    private fun interpretTestFile(fileName: String, rootElement: String): Optional<Element> {
        val testFile = TestFileReader.readTestFile(fileName)
        val sut = XmlWeaveInteractor()
        return sut.getElementStructure(testFile, rootElement)
    }

}